Chatkit/Repositories/JoinedRoomsRepository.swift

```diff
@@ -0,0 +1,148 @@
+
+/// A repository which exposes a collection of all rooms joined by the user.
+///
+/// Construct an instance of this class using `Chatkit.createJoinedRoomsRepository(...)`
+///
+/// ## What is provided
+///
+/// The repository exposes a state with a set of `rooms` which presents the rooms that the current user
+/// is a member of.
+///
+/// ## Receiving live updates
+///
+/// In order to be notified when the contents of the `rooms` or the `state` of the connection changes,
+/// implement the `JoinedRoomsRepositoryDelegate` protocol and assign
+/// the `JoinedRoomsRepository.delegate` property.
+///
+/// ## Understanding the `state` of the repository
+///
+/// The `state` property describes the state of the live update connection, either
+///   - `.initializing`: awaiting the initial set of data, or
+///   - `.connected`: updates are flowing live, or
+///   - `.degraded`: updates may be delayed due to network problems, or
+///   - `.closed`: the connection is closed, no further updates available.
+public class JoinedRoomsRepository: JoinedRoomsRepositoryProtocol {
+    
+    // MARK: - Types
+    
+    typealias Dependencies = HasTransformer
+    
+    // MARK: - Properties
+    
+    private let buffer: Buffer
+    private let connectivityMonitor: ConnectivityMonitor
+    private let dependencies: Dependencies
+    
+    private var versionedState: VersionedState?
+    private var connectionState: ConnectionState
+    
+    /// The current state of the repository.
+    public private(set) var state: State {
+        didSet {
+            if state != oldValue {
+                self.delegate?.joinedRoomsRepository(self, didUpdateState: state)
+            }
+        }
+    }
+    
+    /// The object that is notified when the `state` has changed.
+    public weak var delegate: JoinedRoomsRepositoryDelegate?
+    
+    // MARK: - Initializers
+    
+    init(buffer: Buffer, connectivityMonitor: ConnectivityMonitor, dependencies: Dependencies) {
+        self.buffer = buffer
+        self.connectivityMonitor = connectivityMonitor
+        self.dependencies = dependencies
+        
+        self.versionedState = buffer.currentState
+        self.connectionState = connectivityMonitor.connectionState
+        
+        self.state = JoinedRoomsRepository.state(forConnectionState: connectivityMonitor.connectionState,
+                                                 currentVersionedState: buffer.currentState,
+                                                 previousVersionedState: nil,
+                                                 usingTransformer: dependencies.transformer)
+        
+        self.buffer.delegate = self
+        self.connectivityMonitor.delegate = self
+    }
+    
+    // MARK: - Private methods
+    
+    private static func state(forConnectionState connectionState: ConnectionState, currentVersionedState: VersionedState?, previousVersionedState: VersionedState?, usingTransformer transformer: Transformer) -> State {
```

Comment:
I think the label `currentVersionedState` here should just be `versionedState`.  "current" implies there is some state being held which isn't the case or something this static func cares about.

Reply:
I can rename this parameter.